waddrmgr/db.go

```diff
@@ -1266,10 +1303,102 @@ func upgradeManager(namespace walletdb.Namespace) error {
 		return managerError(ErrDatabase, str, err)
 	}
 
-	// Upgrade the manager as needed.
-	if version < LatestMgrVersion {
-		// No upgrades yet.
+	return nil
+}
+
+// upgradeManager upgrades the data in the provided manager namespace to newer
+// versions as neeeded.
+func upgradeManager(namespace walletdb.Namespace) error {
+	// Get the current version.
+	var version uint32
+	err := namespace.View(func(tx walletdb.Tx) error {
+		mainBucket := tx.RootBucket().Bucket(mainBucketName)
+		verBytes := mainBucket.Get(mgrVersionName)
+		version = binary.LittleEndian.Uint32(verBytes)
+		return nil
+	})
+	if err != nil {
+		str := "failed to fetch version for update"
+		return managerError(ErrDatabase, str, err)
 	}
 
+	// NOTE: There are currently no upgrades, but this is provided here as a
+	// template for how to properly do upgrades.  Each function to upgrade
+	// to the next version must include serializing the new version as a
+	// part of the same transaction so any failures in upgrades to later
+	// versions won't leave the database in an inconsistent state.  The
+	// putManagerVersion function provides a convenient mechanism for that
+	// purpose.
+	//
+	// Upgrade one version at a time so it is possible to upgrade across
+	// an aribtary number of versions without needing to write a bunch of
+	// additional code to go directly from version X to Y.
+	// if version < 2 {
+	// 	// Upgrade from version 1 to 2.
+	//	if err := upgradeToVersion2(namespace); err != nil {
+	//		return err
+	//	}
+	//
+	//	// The manager is now at version 2.
+	//	version = 2
+	// }
+	// if version < 3 {
+	// 	// Upgrade from version 2 to 3.
+	//	if err := upgradeToVersion3(namespace); err != nil {
+	//		return err
+	//	}
+	//
+	//	// The manager is now at version 3.
+	//	version = 3
+	// }
+
+	if version < 2 {
+		// The manager is now at version 2.
+		version = 2
+
+		// Upgrade from version 1 to 2.
+		if err := upgradeToVersion2(namespace, version); err != nil {
+			// Rollback manager to version 1
+			version = 1
+			return err
+		}
+	}
+
+	// Ensure the manager is upraded to the latest version.  This check is
+	// to intentionally causes a failure if the manager version is updated
+	// without writing code to handle the upgrade.
+	if version < latestMgrVersion {
+		str := fmt.Sprintf("the latest manager version is %d, but the "+
+			"current version after upgrades is only %d",
+			latestMgrVersion, version)
+		return managerError(ErrUpgrade, str, nil)
+	}
+
+	return nil
+}
+
+// upgradeToVersion2 upgrades the database from version 1 to version 2
+// 'usedAddrBucketName' a bucket for storing addrs flagged as marked is
+// initialized and it will be updated on the next rescan.
+func upgradeToVersion2(namespace walletdb.Namespace, currentMgrVersion uint32) error {
```

Comment:
The function's sole purpose to upgrade to version 2, so it shouldn't take a version.  I notice the code below is setting it to 2 manually (as expected).